### PR TITLE
nrf5340_audio: refactor power module

### PR DESCRIPTION
**Pull request by gmarull**
This PR refactors the nrf5340_audio power module by using INA23X sensor.

### CODE REVIEW

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [inline applications/nrf5340_audio/src/modules/power_meas.c:2]: Year change.

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [inline applications/nrf5340_audio/src/modules/power_meas.c:21]: Nitpick: Comments should be used where they add readability, and in this struct the comments almost repeat the variable names. This will also increase the maintenance workload.

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [inline applications/nrf5340_audio/src/modules/power_meas.c:37]: IMO: Some comments not required.

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [inline applications/nrf5340_audio/src/modules/power_meas.c:99]: Is the unsigned "U" required?

Comment (koffes) [reply applications/nrf5340_audio/src/modules/power_meas.c:99]: For loops tend to use `int` in Zephyr.

**Review by koffes — CHANGES_REQUESTED** [review]
Thanks for the contribution! This solution is a lot neater. We need to sync with Alexander on testing before I can approve.

**Comment by greg-fer** [comment]
@alexsven , should we add a changelog entry for this stating that the application is not using a custom implementation of the INA driver anymore, and instead uses the upstream one?

**Comment by alexsven** [comment]
> @alexsven , should we add a changelog entry for this stating that the application is not using a custom implementation of the INA driver anymore, and instead uses the upstream one?

Yes, that sounds reasonable

**Review by alexsven — APPROVED** [review]
Tested locally, seems to give reasonable values, but I haven't measured externally to compare the two different implementations. Very nice work!

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c:107]: If we follow Zephyr coding guidelines, we should be using essentially boolean expressions, e.g. `ret < 0` or `ret != 0`, but not `ret`. Since error is specified only in case ret < 0, I tend to prefer that choice (> 0 is not specified as an error or a valid value in many APIs).

**Review by koffes — CHANGES_REQUESTED** [review]
Some nitpick comments are unresolved, but the main issue still lies with verifying the actual readouts. @alexsven nice that you have tested, but we still need to run at least a PPK in series to confirm.

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c]: done

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c]: removed

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c]: removed

**Comment by gmarull** [comment]
Please also approve https://github.com/nrfconnect/sdk-zephyr/pull/972

**Review by greg-fer — COMMENTED** [review]
@gmarull , can you add a changelog entry for this change under Applications > nRF5340 Audio?

**Comment by gmarull** [comment]
> @gmarull , can you add a changelog entry for this change under Applications > nRF5340 Audio?

done

**Comment by gmarull** [comment]
This PR should now be ready, please, make any necessary measurements.

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [inline applications/nrf5340_audio/src/modules/power_meas.c:102]: I would like to share a thought, which is not necessarily a blocker:
 
The way this module has been rewritten, a timer keeps track of when to read the INAs.
This is not ideal, as the timers in the INAs themselves should keep track of this. An average measurement is started, then INA x will assert the alert pin. We should then trigger an interrupt here, read which INA has finished the measurement and then read out the values.

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (koffes) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: If, on the other hand, this diver is only configured for single-triggering measurements (as is done in `sensor_sample_fetch`):

This seems to void the INA function of averaging over N measurements. This reduces the accuracy and gives more I2C traffic.
Also, how do we know that the driver know that the INA has finished when we call `sensor_channel_get`? I have not dug into the lower level drivers yet, but it may be that if the measurement duration is long, we can risk reading out values before the measurement has finsihed.

Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: note: alert was not working at all, at least on my board (I could observe warnings). I actually get similar readings with current setup.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts
Comment (erikrobstad) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts:8]: This file is called ina230.h, but the board is using INA231.
Maybe change name to something like ina23x.h?

File: west.yml
Comment (erikrobstad) [inline west.yml:55]: I guess this will be removed when a rebase is done?

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts
Comment (erikrobstad) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts:71]: the "ina230" name is from when the driver was written with INA230 in mind right? Should be changed to ina23x

Comment (erikrobstad) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts:71]: Same with INA230_CONFIG

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts
Comment (gmarull) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_common.dts:8]: upstream driver is compatible with both, but they named it 230. We can modify it upstream to make things more clear, but should not block this PR.

File: west.yml
Comment (gmarull) [reply west.yml:55]: No, this is needed. We're pulling upstream changes.

Comment (rlubos) [reply west.yml:55]: SHA was updated in https://github.com/nrfconnect/sdk-nrf/pull/9073 to a more recent revision, please drop this change.

Comment (gmarull) [reply west.yml]: done

File: doc/nrf/releases/release-notes-changelog.rst
Comment (greg-fer) [inline doc/nrf/releases/release-notes-changelog.rst:153]:
```suggestion
  * Power module has been re-factored so that it uses upstream Zephyr INA23X sensor driver.
```
We follow "one sentence per line" rule for the NCS docs.

**Review by greg-fer — APPROVED** [review]
Minor nit about line formatting. Approving changelog entry.

**Review by koffes — CHANGES_REQUESTED** [review]
Great! Just one issue left to discuss/alter

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: Note that I have copied all the existing configurations, do you suggest to increase polling interval or reduce averaging time?

Comment (koffes) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: The configs are fine. What I suggest is that this line: `k_msleep(CONFIG_POWER_MEAS_INTERVAL_MS);` is replaced by the actual configuration set in DTS

Comment (koffes) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: To elaborate: `config POWER_MEAS_INTERVAL_MS` only affects the prints and not the actual power meas interval. This KConfig can be removed and rather use pseudo:
`k_msleep((ina230.config.conv_time_curr * + ina230.config.conv_time_voltage)*ina230.config.avg_mode),`

File: doc/nrf/releases/release-notes-changelog.rst
Comment (gmarull) [reply doc/nrf/releases/release-notes-changelog.rst:153]: fixed

File: applications/nrf5340_audio/src/modules/power_meas.c
Comment (gmarull) [reply applications/nrf5340_audio/src/modules/power_meas.c:102]: Updated to 8500ms, and added details on how the value should be configured.

**Review by koffes — APPROVED** [review]

**Pull request merged**
